nestkernel/stopwatch.h

```diff
@@ -0,0 +1,505 @@
+  }
+#endif
+}
+
+template < clockid_t clock_type >
+inline void
+StopwatchBase< clock_type >::stop()
+{
+#ifndef DISABLE_TIMING
+  if ( isRunning() )
+  {
+    _end = get_current_time(); // invariant: _end >= _beg
+    _running = false;          // we stopped running
+  }
+#endif
+}
+
+template < clockid_t clock_type >
+inline bool
+StopwatchBase< clock_type >::isRunning() const
+{
+#ifndef DISABLE_TIMING
+  return _running;
+#else
+  return false;
+#endif
+}
+
+template < clockid_t clock_type >
+inline double
+StopwatchBase< clock_type >::elapsed( timeunit_t timeunit ) const
+{
+#ifndef DISABLE_TIMING
+  size_t time_elapsed;
+  if ( isRunning() )
+  {
+    // get intermediate elapsed time; do not change _end, to be const
+    time_elapsed = get_current_time() - _beg + _prev_elapsed;
+  }
+  else
+  {
+    // stopped before, get time of current measurement + last measurements
+    time_elapsed = _end - _beg + _prev_elapsed;
+  }
+  return static_cast< double >( time_elapsed ) / timeunit;
+#else
+  return 0.;
+#endif
+}
+
+template < clockid_t clock_type >
+inline void
+StopwatchBase< clock_type >::reset()
+{
+#ifndef DISABLE_TIMING
+  _beg = 0; // invariant: _end >= _beg
+  _end = 0;
+  _prev_elapsed = 0; // erase all prev. measurements
+  _running = false;  // of course not running.
+#endif
+}
+
+template < clockid_t clock_type >
+inline void
+StopwatchBase< clock_type >::print( const char* msg, timeunit_t timeunit, std::ostream& os ) const
+{
+#ifndef DISABLE_TIMING
+  double e = elapsed( timeunit );
+  os << msg << e;
+  switch ( timeunit )
+  {
+  case MICROSEC:
+    os << " microsec.";
+    break;
+  case MILLISEC:
+    os << " millisec.";
+    break;
+  case SECONDS:
+    os << " sec.";
+    break;
+  case MINUTES:
+    os << " min.";
+    break;
+  case HOURS:
+    os << " h.";
+    break;
+  case DAYS:
+    os << " days.";
+    break;
+  }
+#ifdef DEBUG
+  os << " (running: " << ( _running ? "true" : "false" ) << ", begin: " << _beg << ", end: " << _end
+     << ", diff: " << ( _end - _beg ) << ", prev: " << _prev_elapsed << ")";
+#endif
+  os << std::endl;
+#endif
+}
+
+template < clockid_t clock_type >
+inline size_t
+StopwatchBase< clock_type >::get_current_time()
+{
+  // We use a monotonic timer to make sure the stopwatch is not influenced by time jumps (e.g. summer/winter time).
+  timespec now;
+  clock_gettime( clock_type, &now );
+  return now.tv_nsec + now.tv_sec * timeunit_t::SECONDS;
+}
+
+template < clockid_t clock_type >
+inline std::ostream&
+operator<<( std::ostream& os, const StopwatchBase< clock_type >& stopwatch )
+{
+  stopwatch.print( "", timeunit_t::SECONDS, os );
+  return os;
+}
+
+}
+
+enum StopwatchVerbosity
+{
+  Normal,  //<! Always measure stopwatch
+  Detailed //<! Only measure if detailed stopwatches are activated
+};
+
+enum StopwatchType
+{
+  MasterOnly, //<! Only the master thread owns a stopwatch
+  Threaded    //<! Every thread measures an individual stopwatch
+};
```

Comment:
Also here I think we could use more expressive names:
```suggestion
enum StopwatchParallelism
{
  MasterOnly, //<! Only the master thread owns a stopwatch
  PerThread   //<! Every thread owns an individual stopwatch
};
```